### DISCUSSION TITLE
如何让在 docker-compose 中运行的 wget 命令使用 Xray 的透明代理。

### DISCUSSION THREAD

**Discussion by zw963** [XTLS/Xray-core#1749] [General]
大家好。

目前我在使用一个老电脑做旁路由，使用 Xray 架设透明代理，用了好久了， 一切工作的还算好，[这里](https://github.com/zw963/asuswrt-merlin-v2ray/blob/master/router/opt/etc/apply_iptables_rule.sh#L70-L113) 是我使用的 iptables 设置。

但是一直有个问题困扰我，没有解决。

就是当我使用 docker 或者 docker-compose 的时候，在 docker 中跑 wget 或 curl 的时候，总是因为墙的问题失败。

例如，下面的代码当 docker-compose build 的时候，使用 curl 访问 github 错误提示
```
/app # curl -fsSLo- https://raw.githubusercontent.com/zw963/sentry/master/install.cr | crystal eval
🤖  Fetching sentry files...Unhandled exception: Error connecting to 'raw.githubusercontent.com:443': Address not available (Socket::ConnectError)
  from /usr/share/crystal/src/socket/addrinfo.cr:67:17 in 'initialize'
  from /usr/share/crystal/src/socket/tcp_socket.cr:27:3 in 'initialize'
  from /usr/share/crystal/src/socket/tcp_socket.cr:27:3 in 'new'
  from /usr/share/crystal/src/http/client.cr:795:10 in 'io'
  from /usr/share/crystal/src/http/client.cr:676:19 in 'send_request'
  from /usr/share/crystal/src/http/client.cr:607:5 in 'exec_internal_single'
  from /usr/share/crystal/src/http/client.cr:590:18 in 'exec_internal'
  from /usr/share/crystal/src/http/client.cr:583:7 in 'exec'
  from /usr/share/crystal/src/http/client.cr:719:5 in 'exec'
  from /usr/share/crystal/src/http/client.cr:751:7 in 'exec'
  from /usr/share/crystal/src/http/client.cr:408:3 in 'get'
  from eval:9:1 in '__crystal_main'
  from /usr/share/crystal/src/crystal/main.cr:115:5 in 'main_user_code'
  from /usr/share/crystal/src/crystal/main.cr:101:7 in 'main'
  from /usr/share/crystal/src/crystal/main.cr:127:3 in 'main'
  from /lib/ld-musl-x86_64.so.1 in '??'
/app # 
```

但是，在本地主机直接跑(透明代理)却工作的很好， 即: 透明代理使用的 iptables rule 在 container 里面不工作。

请问，这种问题该如何解决? 谢谢！

**Comment by cross-hello** [comment]
Say if your local http/https listening port is 8080:
Set wget agent configuration in docker:
`echo -e "use_proxy=yes\nhttp_proxy=127.0.0.1:8080\nhttps_proxy=127.0.0.1:8080" >>~/.wgetrc `

And `echo -e
 "export http_proxy=127.0.0.1:8080\nexport https_proxy=127.0.0.1:8080\n">>~/.bashrc && source ~/.bashrc` for curl.

**Reply by zw963** [reply]
多谢，这不就相当于开全局代理了嘛，我期望的解决方案是，如何写 iptables rule 让 container 内的 wget/curl 也可以和 container 外一样，自动透明代理。

**Reply by cross-hello** [reply]
It is possible to [forward fixed port to another using redsocks](https://medium.datadriveninvestor.com/how-to-transparently-use-a-proxy-with-any-application-docker-using-iptables-and-redsocks-b8301ddc4e1e).

But for float port like curl and wget don't find a proper solution.

Beside, another workaround is directly copy script into docker.

**Reply by zw963** [reply]
我找到了一个 workaround, 我启动了一个 V2RayA, 并且开启了全局代理，现在在容器内使用 curl 访问 raw.githubusercontent.com 可以工作了，虽然还是没解决我的问题，但好歹能用了，多谢！

**Reply by cross-hello** [reply]
🗡️ Various workouts could deal with, and for the few problems you have spent nearly one day.
